README.md

```diff
@@ -1,106 +1,193 @@
+
+### Goals
 
-### Goals [or Motivating Use Cases, or Scenarios]
+The primary goal of this API is to allow users to use WASI programs to access key-value stores. It abstracts away the complexity of building stateful services that need to access key-value stores. In other words, it decouples the application code from the specific key-value store APIs. This allows WASI programs to be portable across different key-value store that supports this interface, be it on-prem, in the cloud, or in edge devices.
 
-[What is the end-user need which this project aims to address?]
+The second goal of this API is to abstract away the network stack, allowing applications that need to access key-value stores to not worry about what network protocol is used to access the key-value store.
 
 ### Non-goals
 
-[If there are "adjacent" goals which may appear to be in scope but aren't, enumerate them here. This section may be fleshed out as your design progresses and you encounter necessary technical and other trade-offs.]
+- Cover all application requirements for a key-value store. Instead, it focuses on the most common use cases. It allows providers to extend the API to support more use cases.
+- Transactional semantics.
+- Data consistency.
+- Data replication.
+- Data sharding.
+- Cluster management.
+- Monitoring
 
 ### API walk-through
 
-[Walk through of how someone would use this API.]
+#### Use case 1
 
-#### [Use case 1]
+Imagine you have an HTTP handler that needs to persist some data to a key-value store. The handler needs to be able to retrieve, delete, and update the data in the key-value store. The following Rust code shows how you can use the WASI Key-Value Store API to in the handler.
 
-[Provide example code snippets and diagrams explaining how the API would be used to solve the given problem]
+```rust
+let value = b"value1";
+kv1::create_or_update("key", value)?;
+let val = kv1::get("key")?;
+assert_eq!(val, value);
 
-#### [Use case 2]
-
-[etc.]
+kv1::delete("key")?;
+```
 
 ### Detailed design discussion
 
-[This section should mostly refer to the .wit.md file that specifies the API. This section is for any discussion of the choices made in the API which don't make sense to document in the spec file itself.]
+```go
+interface "wasi:kv/data/crud" {
+  use { Error, payload, key } from "wasi:kv/types"
+
+  get: func(key: key) -> result<payload, Error>
+
+  create-or-update: func(key: key, value: stream<u8>) -> result<_, Error>
 
-#### [Tricky design choice #1]
+  delete: func(key: key) -> result<_, Error>
+}
 
-[Talk through the tradeoffs in coming to the specific design point you want to make.]
+interface "wasi:kv/data/increment" {
+  use { Error, key } from "wasi:kv/types"
+  
+  increment: func(key: key) -> result<u64, Error>
+}
 
+interface "wasi:kv/data/bulk-get" {
```

Comment:
It might be easier if we have an interface called `wasi:kv/data/bulk` which has the `get` and `put` operations (this is an aside, but within an interface we shouldn't need to prefix the operations with an interface, so `bulk-get` doesn't need the bulk and should just be `get` imho).

Realistically, I think a kv implementation either supports "bulk operations" or not... I've rarely, if ever, seen a data store support bulk reads without bulk writes or the reverse

Reply:
e.g. if you have a `get-many` inside the `atomic` interface, you know that the result set will not be alterable during the query, whereas `get-many` inside the basic interface would imply no such guarantee.

Reply:
I've changed bulk-get and bulk-set to get-many and set-many. To make it clear to distinguish atomic operation or non-atomic operation, I've also created a new interface "wasi:kv/data/transaction" for atomic `bulk-get` and `bulk-set`.